test/unit/math/test_ad.hpp

```diff
@@ -1,22 +1,365 @@
 #ifndef TEST_UNIT_MATH_TEST_AD_HPP
 #define TEST_UNIT_MATH_TEST_AD_HPP
 
-#include <stan/math.hpp>
+#include <stan/math/mix/mat.hpp>
 #include <test/unit/math/util.hpp>
-#include <test/unit/math/mix/mat/util/autodiff_tester.hpp>
 #include <gtest/gtest.h>
+#include <algorithm>
 #include <string>
 #include <vector>
 
 namespace stan {
 namespace test {
 
+/**
+ * Return true if the specified value is finite.
+ *
+ * @param x value to test
+ * @return true if value is finite
+ */
+bool is_finite(double x) {
+  return !stan::math::is_inf(x) && !stan::math::is_nan(x);
+}
+
+/**
+ * Return true if all of the elements in the container are finite
+ *
+ * @tparam T scalar type
+ * @tparam R row type
+ * @tparam C col type
+ * @param x container to test
+ * @return true if all container values are finite
+ */
+template <typename T, int R, int C>
+bool is_finite(const Eigen::Matrix<T, R, C>& x) {
+  for (int i = 0; i < x.size(); ++i)
+    if (!is_finite(x(i)))
+      return false;
+  return true;
+}
+
+/**
+ * Return true if all of the elements in the container are finite
+ *
+ * @tparam T contained type
+ * @param x container to test
+ * @return true if all container values are finite
+ */
+template <typename T>
+bool is_finite(const std::vector<T>& x) {
+  for (size_t i = 0; i < x.size(); ++i)
+    if (!is_finite(x[i]))
+      return false;
+  return true;
+}
+
+/**
+ * Test that the specified values are within the specified tolerance
+ * on relative error, and if not, fail the embedded google test.
+ *
+ * <p>Relative error is defined to be the error `u - v` rescaled by the
+ * average absolute value,
+ * `rel_err(u, v) = (u - v) / (0.5 * (abs(u) * + abs(v))).`
+ *
+ * <p>If both `u` and `v` are zero, the error is defined to be zero.
+ * If only one of `u` or `v` is zero, the above definition always
+ * yields 1.  This is not a useful test, so we instead test absolute
+ * error in the case that one of the values is zero.
+ *
+ * @tparam T1 type of first argument
+ * @tparam T2 type of second argument
+ * @param msg failure message for reporting
+ * @param x1 first argument
+ * @param x2 second argument
+ * @param tol relative tolerance
+ */
+template <typename T1, typename T2>
+void expect_near_relative(const std::string& msg, const T1& x1, const T2& x2,
+                          double tol = 1e-8) {
+  using stan::math::fabs;
+  if (x1 == 0 && x2 == 0)
+    return;
```

Comment:
Because if one is zero, the other should be near, but that can't really be absolute tolerance vs. truth as truth is zero.  So there's a separate EXPECT_NEAR case for `(x1 == 0 || x2 == 0)`.  I'll add doc.